plugins/transport-nio/src/main/java/org/elasticsearch/http/nio/HttpReadWriteHandler.java

```diff
@@ -152,4 +174,18 @@ private void handleRequest(Object msg) {
             request.release();
         }
     }
+
+    private void maybeReadTimeout() {
+        if (requestSinceReadTimeoutTrigger == false && inFlightRequests == 0) {
+            String message = "http read timeout after " + TimeValue.nsecToMSec(readTimeoutNanos) + "ms";
+            transport.onException(nioHttpChannel, new HttpReadTimeoutException(message));
+        } else {
+            requestSinceReadTimeoutTrigger = false;
+            scheduleReadTimeout();
+        }
+    }
+
+    private void scheduleReadTimeout() {
+        taskScheduler.scheduleAtRelativeTime(this::maybeReadTimeout, System.nanoTime() + readTimeoutNanos);
```

Comment:
Maybe pass a `LongSupplier` in the constructor and use  `threadPool::relativeTimeInMillis` for calculations here since we have a `ThreadPool` available in `org.elasticsearch.http.nio.NioHttpServerTransport.HttpChannelFactory#createChannel` anyway?
Then we could also simplify the test in `testReadTimeout` quite a bit and get rid of a bunch of mocking since we can just use the real `TaskScheduler` when we control the clock?

Reply:
++